subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -1110,376 +637,462 @@ static void fota_evt_handler(struct azure_fota_event *fota_evt)
-			return -ENOMEM;
-		}
-
+	switch (state) {
+	case AZURE_IOT_HUB_DPS_REG_STATUS_NOT_STARTED:
+		LOG_DBG("AZURE_IOT_HUB_DPS_REG_STATUS_NOT_STARTED");
 		break;
-	}
-	case AZURE_IOT_HUB_TOPIC_TWIN_REPORTED:
-		len = snprintk(topic, sizeof(topic),
-				TOPIC_TWIN_REPORT, k_uptime_get_32());
-		if ((len < 0) || (len > sizeof(topic))) {
-			LOG_ERR("Failed to create twin report topic");
-			return -ENOMEM;
-		}
+	case AZURE_IOT_HUB_DPS_REG_STATUS_ASSIGNED:
+		LOG_DBG("AZURE_IOT_HUB_DPS_REG_STATUS_ASSIGNED");
+		k_sem_give(&dps_sem);
 		break;
-	case AZURE_IOT_HUB_TOPIC_TWIN_REQUEST:
-		len = snprintk(topic, sizeof(topic),
-			       TOPIC_TWIN_REQUEST, k_uptime_get_32());
-		if ((len < 0) || (len > sizeof(topic))) {
-			LOG_ERR("Failed to create device twin request");
-			return -ENOMEM;
-		}
+	case AZURE_IOT_HUB_DPS_REG_STATUS_FAILED:
+		LOG_DBG("AZURE_IOT_HUB_DPS_REG_STATUS_FAILED");
+		/* TODO: In case of failure, _start() will still wait for the sampehore to time out.
+		 *	 This can be improved to give feedbck to the app sooner.
+		 */
+		break;
+	case AZURE_IOT_HUB_DPS_REG_STATUS_ASSIGNING:
+		LOG_DBG("AZURE_IOT_HUB_DPS_REG_STATUS_ASSIGNING");
 		break;
 	default:
-		LOG_ERR("No topic specified");
-		return -ENOMSG;
+		LOG_WRN("Unhandled DPS state: %d", state);
+		break;
 	}
-
-	param.message.topic.topic.size = len;
-	param.message.topic.topic.utf8 = topic;
-
-	LOG_DBG("Publishing to topic: %s",
-		log_strdup(param.message.topic.topic.utf8));
-
-	return mqtt_publish(&client, &param);
 }
+#endif /* IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS) */
 
-int azure_iot_hub_disconnect(void)
+/* Public interface */
+
+int azure_iot_hub_init(const azure_iot_hub_evt_handler_t event_handler)
 {
-	int err;
+	if (!iot_hub_state_verify(STATE_UNINIT) && !iot_hub_state_verify(STATE_DISCONNECTED)) {
+		LOG_WRN("Azure IoT Hub cannot be re-initialized in this state: %s",
+			state_name_get(iot_hub_state_get()));
+		return -EALREADY;
+	}
 
-	if (!connection_state_verify(STATE_CONNECTED)) {
-		LOG_WRN("Azure IoT Hub is not connected");
-		return -ENOTCONN;
+	if (event_handler == NULL) {
+		LOG_ERR("Event handler must be provided");
+		return -EINVAL;
 	}
 
-	connection_state_set(STATE_DISCONNECTING);
+#if IS_ENABLED(CONFIG_AZURE_FOTA)
+	int err = azure_fota_init(fota_evt_handler);
 
-	err = mqtt_disconnect(&client);
 	if (err) {
-		LOG_ERR("Failed to disconnect MQTT client, error: %d", err);
-		connection_state_set(STATE_INIT);
+		LOG_ERR("Failed to initialize Azure FOTA, error: %d", err);
 		return err;
 	}
 
-	/* The MQTT library only propagates the MQTT_DISCONNECT event
-	 * if the call to mqtt_disconnect() is successful. In that case the
-	 * setting of STATE_INIT is carried out in the mqtt_evt_handler.
-	 */
+	LOG_DBG("Azure FOTA initialized");
+#endif
+
+	evt_handler = event_handler;
+
+	iot_hub_state_set(STATE_DISCONNECTED);
 
 	return 0;
 }
 
-int azure_iot_hub_connect(void)
+int azure_iot_hub_connect(const struct azure_iot_hub_config *config)
```

Comment:
This function is over 150 lines long. Would be nice if parts of it could be extraced into their own functions to improve readability.